From 900543395458afce0989b77620a959089057498d Mon Sep 17 00:00:00 2001
From: Huynh Thanh Hung <hung.huynh.wz@renesas.com>
Date: Thu, 26 Nov 2020 18:22:12 +0700
Subject: [PATCH] plat: renesas: Implement SiP Service Call

This patch will add SiP Service Call feature.
This SiP Service Call just support to get ECC MODE in current.

Signed-off-by: Huynh Thanh Hung <hung.huynh.wz@renesas.com>
---
 plat/renesas/rzg/include/rzg_sip_svc.h | 22 +++++++++++
 plat/renesas/rzg/platform.mk           |  1 +
 plat/renesas/rzg/rzg_sip_svc.c         | 51 ++++++++++++++++++++++++++
 3 files changed, 74 insertions(+)
 create mode 100644 plat/renesas/rzg/include/rzg_sip_svc.h
 create mode 100644 plat/renesas/rzg/rzg_sip_svc.c

diff --git a/plat/renesas/rzg/include/rzg_sip_svc.h b/plat/renesas/rzg/include/rzg_sip_svc.h
new file mode 100644
index 000000000..a6d5dc8af
--- /dev/null
+++ b/plat/renesas/rzg/include/rzg_sip_svc.h
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2020, Renesas Electronics Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef RZG_SIP_SVC_H__
+#define RZG_SIP_SVC_H__
+
+/* Function ID to get ECC mode */
+#define RZG_SIP_SVC_GET_ECC_MODE	U(0x8200000F)
+
+/* Rcar SiP Service Calls version numbers */
+#define RZG_SIP_SVC_VERSION_MAJOR	U(0x0)
+#define RZG_SIP_SVC_VERSION_MINOR	U(0x0)
+
+#define RZG_SMC_RET_SUCCESS		0
+#define RZG_SMC_RET_EFAILED		-2
+#define RZG_SMC_RET_PMIC_DISABLE	-3
+
+
+#endif /* RZG_SIP_SVC_H__ */
diff --git a/plat/renesas/rzg/platform.mk b/plat/renesas/rzg/platform.mk
index 3504bfeca..8834492e4 100644
--- a/plat/renesas/rzg/platform.mk
+++ b/plat/renesas/rzg/platform.mk
@@ -372,6 +372,7 @@ BL31_SOURCES	+=	${RZG_GIC_SOURCES}				\
 			plat/renesas/rzg/aarch64/platform_common.c	\
 			plat/renesas/rzg/bl31_plat_setup.c		\
 			plat/renesas/rzg/plat_pm.c			\
+			plat/renesas/rzg/rzg_sip_svc.c			\
 			drivers/renesas/rzg/console/rzg_console.S	\
 			drivers/renesas/rzg/console/rzg_printf.c	\
 			drivers/renesas/rzg/delay/micro_delay.c		\
diff --git a/plat/renesas/rzg/rzg_sip_svc.c b/plat/renesas/rzg/rzg_sip_svc.c
new file mode 100644
index 000000000..29f2c928c
--- /dev/null
+++ b/plat/renesas/rzg/rzg_sip_svc.c
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2016-2017, ARM Limited and Contributors. All rights reserved.
+ * Copyright (c) 2020, Renesas Electronics Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <common/runtime_svc.h>
+#include <common/debug.h>
+#include <smccc_helpers.h>
+#include <string.h>
+#include <arch_helpers.h>
+#include "rzg_sip_svc.h"
+
+/*
+ * This function handles RZG defined SiP Calls
+ */
+static uintptr_t rzg_sip_handler(unsigned int smc_fid,
+			u_register_t x1,
+			u_register_t x2,
+			u_register_t x3,
+			u_register_t x4,
+			void *cookie,
+			void *handle,
+			u_register_t flags)
+{
+	int32_t		arg[2];
+
+	switch (smc_fid) {
+		case RZG_SIP_SVC_GET_ECC_MODE:
+			arg[0] = RZG_DRAM_ECC;
+			arg[1] = RZG_DRAM_ECC_FULL;
+			SMC_RET2(handle, arg[0], arg[1]);
+
+		default:
+			WARN("Unimplemented RZG SiP Service Call: ");
+			WARN("0x%x \n", smc_fid);
+			SMC_RET1(handle, SMC_UNK);
+	}
+}
+
+
+/* Define a runtime service descriptor for fast SMC calls */
+DECLARE_RT_SVC(
+	rzg_sip_svc,
+	OEN_SIP_START,
+	OEN_SIP_END,
+	SMC_TYPE_FAST,
+	NULL,
+	rzg_sip_handler
+);
-- 
2.17.1

